File: practices/chapter12_recursive/python/ch_12_practices.py
import math

def recursive_cal_arr_length(arr=[], prev=0):
  if not arr:
    return prev
  arr.pop()
  return recursive_cal_arr_length(arr, prev + 1)

def recursive_find_max(data_list=[], max_now=math.inf*-1):
  if not data_list:
    return max_now
  data_now = data_list.pop()

  if data_now > max_now:
    return recursive_find_max(data_list, data_now)
  else:
    return recursive_find_max(data_list, max_now)

def accumulate_1(n):
  if n == 0:
    return 0
  res = (1 / n) + accumulate_1(n - 1)
  print('accumulate_1({}) = {}'.format(n, res))
  return res

def accumulate_2(n):
  if n == 0:
    return 0
  res = (n / (n + 1)) + accumulate_2(n - 1)
  print('accumulate_2({}) = {}'.format(n, res))
  return res

def main():
  data_list = [1, 10, 4, 6]
  print('Length: ', recursive_cal_arr_length(list(data_list)))
  print('Max: ', recursive_find_max(list(data_list)))

def accumulate_sample():
  n = int(input('Input n: '))
  accumulate_1(n)
  accumulate_2(n)

if __name__ == "__main__":
    # main()
    accumulate_sample()

File: practices/chapter12_recursive/js/ch_12_practices.js
const {
  rl,
} = require('../../lib/cmdReadline')

function recursiveGetArrLength(dataList=[]) {
  let _dataList = [...dataList]
  const innerFn = (prev) => {
    if (_dataList.length === 0) {
      return prev
    }
    _dataList.pop()
    return innerFn(prev + 1)
  }
  return innerFn(0)
}

function recursiveFindMax(dataList=[]) {
  let _dataList = [...dataList]
  if (_dataList.length === 0) {
    return undefined
  }

  const innerFindMax = (maxNow) => {
    if(_dataList.length === 0) {
      return maxNow
    }
    const data = _dataList.pop()
    if(data > maxNow) {
      return innerFindMax(data)
    }
    return innerFindMax(maxNow)
  }

  return innerFindMax(_dataList[0])
}

function accumulate1(n=1) {
  if(n === 0) {
    return 0
  }
  const res = (1 / n) + accumulate1(n - 1)
  console.log(`accumulate1(${n}) = ${res}`)
  return res
}

function accumulate2(n=1) {
  if(n === 0) {
    return 0
  }
  const res = (n / (n + 1)) + accumulate2(n - 1)
  console.log(`accumulate2(${n}) = ${res}`)
  return res
}

function main() {
  const dataList = [1, 3, 20, 4, 6, 30]
  console.log('Get array length(recursively): ', recursiveGetArrLength(dataList))
  console.log('Get max in array(recursively): ', recursiveFindMax(dataList))
}

function accumulateSample() {
  rl.question('Input n: ', ans => {
    let n = Number(ans)
    accumulate1(n)
    accumulate2(n)

    rl.close()
  })
}

// main()
accumulateSample()